### ISSUE TITLE
Cubical Agda & non-injectivity of Glue

### ISSUE THREAD

**Issue by AndrasKovacs** [agda/agda#5755]
I'm trying to catch up on the rules used in modern cubical TT implementations. For Agda, I looked at http://www.cse.chalmers.se/~simonhu/misc/hcomp.pdf, which is AFAIK the only online documentation on computation rules on `Glue`. The [source code](https://hackage.haskell.org/package/Agda-2.6.2.1/docs/src/Agda.TypeChecking.Primitive.Cubical.html#compGlue) appears to largely follow this.

Now, looking at 2.1 and 3.6 in Huber's note, it seems to me that 3.6. is incorrect. For `transpⁱ A φ u₀` we assume `i : I, φ ⊢ A ≡ A[0/i]`. This implies in particular that `φ ⊢ A[1/i] ≡ A[0/i]`, since definitional equality is stable under substitution.

Now, going to the definition of `transpⁱ (Glue [φ ↦ (T, w)] A) ψ u₀`, it is written that "the above data is constant when restricted to `ψ`". This does not follow. We have by assumption

    i : I, Ψ ⊢ Glue [φ ↦ (T, w)] A ≡ (Glue [φ ↦ (T, w)] A)[0/i]

therefore

    i : I, Ψ ⊢ Glue [φ ↦ (T, w)] A ≡ Glue [φ[0/i] ↦ (T[0/i], w[0/i])] A[0/i])

`Glue` is not definitionally injective, so we can't conclude constancy of all component types/terms. For example, `α ⊢ Glue [α ↦ (T, w)] A ≡ T ≡ Glue [α ↦ (T, w')] A'`, for `A ≢ A'` and `w ≢  w'`.

This already looks a bit sketchy to me when we specify `transp` rules for type formers which are actually injective. For example, `transpⁱ ((a : A) → B) φ` relies on injectivity of `Π`, but when we are in the process of giving an algebraic specification for the theory, we don't yet know this! Injectivity of `Π` is provable by normalization. In the algebraic specification, we would have to include extra assumptions for the constancy of `A` and `B` under `φ`, which become redundant after we prove `Π` injectivity.  However, the extra assumptions may complicate or break the (hypothetical) normalization proof.

**Comment by AndrasKovacs** [comment]
Here's a more concrete example. 

```agda
{-# OPTIONS --cubical --show-irrelevant #-}

module TranspGlue where

open import Cubical.Foundations.Everything
open import Agda.Builtin.Cubical.Path
open import Agda.Builtin.Cubical.Glue

module Mod (A₀ A₁ : Type)(A≡ : A₀ ≡ A₁) (α : I) (e : ∀ i → A₀ ≃ A≡ i) where

  G : I → Type
  G i = Glue (A≡ i) {α} (λ _ → A₀ , e i)

  -- G is constantly A₀ under α
  constantG : ∀ i → Sub Type α λ _ → A₀
  constantG i = inS (G i)

  -- so transport works
  transpG : G i0 → G i1
  transpG = transp G α

  -- The computation rule assumes that A≡ i is definitionally equal to A≡ i0 under α.
  -- Which is not the case, so the definition below computes to an ill-typed term.
  f : G i0 → A₁
  f g₀ = unglue α (transpG g₀)
```

At the final `f` definition, I can print out the normal form, and try to typecheck it, but this gives multiple errors which are pretty much impossible to interpret, and which are not necessarily related to the problem that I described. 

Anyway, I don't think it's essential to try to check the normal form. It should be enough for my purpose to look at the following line.

https://github.com/agda/agda/blob/e2eb867be0353cec5ad1a3edf0ef209b3ed6aef4/src/full/Agda/TypeChecking/Primitive/Cubical.hs#L583

And note that in the context of my code, `a0` above is `unglue g₀ : A₀`, which is ill-typed inside the `gcomp`, because it should have type `A≡ i` instead (`bA <@> i` in the Agda source). The constancy constraint `α` (`psi` in the source) is supposed to ensure that the two are equal, but they aren't.

I find this rather worrying. Preliminary thoughts on possible fixes.
- Rewrite `transp` for `Glue` without relying on the constancy of the `Glue` sub-components. It seems to me that this is not possible. At least we have to send the `unglue` part from `A i0` to `A i1` in a way that's the identity function under the constancy constraint. But without constancy for `A`, there is no operation that satisfies this.
- Drop the constancy constraints altogether from `transp`. I'm not yet too familiar with the system, so I don't know what would be broken by this. At least, since CCHM had univalence without constrained `transp`, univalence should be still possible, right? 
- Make `Glue` definitionally injective as a type former. This means that `Glue` would not strictly compute to the endpoint type, but instead it would be definitionally isomorphic to it (or something like this). But then how do we get `ua`? 
- Check constancy at "runtime", blocking reduction if things aren't constant enough. I find this very hacky but maybe it's good  as a temporary fix for type soundness.

**Comment by AndrasKovacs** [comment]
After reviewing [CHM](https://arxiv.org/pdf/1802.01170.pdf) and digesting this problem a bit more, "checking constancy during reduction" seems the most practically sensible solution. There's another somewhat sensible design, namely the one in CHM sections 3.1-3.3, where `comp` is the primitive Kan operation for *arbitrary* types, and `hcomp` and `transp` are bundled to specific canonical & injective type formers. I review these options.

#### 1. Checking constancy during reduction

The solution is conceptually simple: when computing `transpⁱ (Glue [α ↦ (T, e)] A) ψ`, only reduce if `α`, `A` and `e` are all constant in `i` under `ψ` (`T` is already known to be constant). This requires performing a conversion check during reduction/evaluation, and of course we have to re-check this "reduction condition" whenever we perform interval substitution.

How sketchy is this? Well, this would be the *only* place, if I'm correct, where reduction depends on conversion checking, so that's one thing. In this sense it's similar to the "regularity" rule. I don't know much about cubical TT semantics, so this rule may or may not be fine there. My intuition says that it's fine, because we can always just simply drop or weaken equations in any semantic model.

Moreover, I think that by using the `transp`-`hcom` split, we have already accepted its sketchiness wholesale, so a bit more of it doesn't matter much. As I mentioned upthread, just by having the `transp` primitive, *every open type former*  must formally require extra constancy conditions in every `transp` computation rule. For example, the `transp` rule for `cons x xs` in type `List A` must assume that `A` is constant under the given constraint. When we define the notion of model (signature) for the theory, we have no injectivity of type formers; that may be just false in a given model. A practical implementation of CHM may omit the constancy checks, because in the syntactic model the type former injectivities can be proven.

How sketchy is this from the perspective of normalization/canonicity proofs? Here I'm really not that familiar with the related recent cubical literature. I do know from NbE for ordinary (non-cubical) type theories, that conversion checking during evaluation is not an issue formally, because we are free to quote values to normal forms at any point.

#### 2. Having primitive `comp`

If for some reason we really don't want to do the previous solution, and want to stick with De Morgan CTT, we can switch to only having `comp` for arbitrary types. Following CHM, we can additionally introduce `hcomp` and `transp` for each HIT, and compute `comp` in terms of those. The issues here are that we lose efficiency for HITs, and it would be a significant change in the Agda codebase.

In conclusion, only option **1.** looks practically realistic to me now.

**Comment by nad** [comment]
Perhaps @Saizan or @mortberg could comment about this.

**Comment by sattlerc** [comment]
I can't speak to the implementation in Cubical Agda, but the incorrect unfolding of the arguments for transp for Glue in http://www.cse.chalmers.se/~simonhu/misc/hcomp.pdf looks like a simple mistake with no consequences. The definition does not depend on it. (We still have Γ, ψ, i : 𝕀, φ ⊢ T = T(i/0).)

Recall that `transp` and `hcomp` are special cases of `comp`.

`comp` for `Glue [φ ↦ (T, w)] A` does not have any strict equalities of types as assumptions and can be defined using `comp` for `T` and `A`. We are interested in the special case of `transp` for `Glue [φ ↦ (T, w)] A`. This turns the use of `comp` for `T` into `transp` for `T`. And `comp` for `A` reduces to `transp` and `hcomp` for `A`. That's how `transp` for `Glue` can be seen to arise.

**Comment by AndrasKovacs** [comment]
@sattlerc yeah, that's what I describe in my previous comment as option 2. It does have consequence in the Agda codebase, because it needs significant refactoring and it appears to be less efficient. For an example for inefficiency, if I do

    transport (λ i → Susp (A i)) north

In the current Agda it computes to `north` right away. But if we only have `comp` as primitive at arbitrary types, the generic `transport` is defined in terms of `comp`, and `comp` at `Susp` type is defined in terms of a recursive `transpSusp` operation and a formal `hcompSusp` (according to Section 3.2 of CHM). In this case `transport (λ i → Susp (A i)) north` reduces to an empty `hcomp` on `north`. In general, transporting a non-indexed HIT constructor would always produce an empty `hcomp`.

The HIT behavior change would also break a lot of existing cubical Agda code, while the option of adding extra checks to `transp Glue` would probably break no code at all.

The extra checks on `transp Glue` also incur performance cost of course, but they can be easily implemented and benchmarked on existing code to see how bad it is. My intuition is that it won't be too bad.

**Comment by sattlerc** [comment]
I was not suggesting to take `comp` as primitive. I was just explaining the high-level perspective of how `transp` for `Glue [φ ↦ (T, w)] A` arises from `transp` for `T` and `transp` and `hcomp` for `A` (using the `comp`-based point of view as given).

None of options 1 and 2 should be needed.

**Comment by AndrasKovacs** [comment]
OK I think I get it. You say that the computation rule should be just

    transpⁱ (Glue [α ↦ (T, f)] A) β g₀ := compGlueⁱ [α ↦ (T, f)] A [β ↦ g₀] g₀

where `compGlueⁱ` is the computation rule for `comp Glue`, and we can use that without needing primitive `comp`. I was fixated on how the original definition breaks without the `A` constancy assumption, but indeed there's a different definition which does not depend on that. 

That's a lot better than my overcomplicated suggestions. The fix is just to rewrite the `transp Glue` definitions.

**Comment by Saizan** [comment]
> That's a lot better than my overcomplicated suggestions. The fix is just to rewrite the `transp Glue` definitions.

As you say though the assumed injectivity is relied on for other type formers, so algebraically we have to revise those rules as well, if we want them to be sound in cubical sets with the usual interpretation of type formers.
Once we do that we would have to see if the current rules are still admissible or not.

There is also a fourth option here: move to some version of cartesian cubical type theory, or at least a theory with something more like `coe` than `transp`. Eventually we want to do something like that to be sound wrt simplicial sets.

**Comment by mortberg** [comment]
Indeed, I agree that @sattlerc 's suggestion is the best short term fix. 

In the long run we should probably consider doing what @Saizan suggests and switch to the cartesian Kan operations, but that's a much bigger change. It's not clear to me if we can do this while keeping the connections and reversals on the interval while still having decidable typechecking as we would have to add "diagonal cofibrations". I have a vague recollection that @Saizan thought this was still going to be fine though?

**Comment by simhu** [comment]
Thanks, @AndrasKovacs, for pointing out this possible problem.  One thing I should clarify about the note: the equalities should be read as premises to the judgmental equalities (so, e.g., `transp` for `(x:A) -> B` requires suitably constant `A` and `B` to compute).  But these premises might be too much of a requirement in the case of `Glue`, and it doesn't seem unlikely that an example like yours can be manufactured which breaks canonicity for the described type theory.  (There seems to be no problem in the semantics though.)

But as @sattlerc mentioned, the premises are not essential, the same algorithm works.  Or, well, *almost* the same algorithm, but fixing a small subtlety at exactly the place @AndrasKovacs mentioned earlier:

https://github.com/agda/agda/blob/e2eb867be0353cec5ad1a3edf0ef209b3ed6aef4/src/full/Agda/TypeChecking/Primitive/Cubical.hs#L583

I have updated the note accordingly here: https://simhu.github.io/misc/hcomp.pdf
The change is in the partial element in the definition of `a1`, should be rather simple to patch (@Saizan, @mortberg? ;D).  I don't know which algorithm `compGlue` is, but fixing `transp` as described seems to me more direct.

But I didn't check the algorithm in the Agda code which deviates in the end (a1', t1') from the one described in the note.  This might be one of the optimizations someone came up with at some point?  I don't recall, but this looks a bit sketchy.

**Comment by Saizan** [comment]
@simhu thanks for chiming in! Have you worked out the cubical sets semantics of transp in e.g. Pi types though?
If we take the standard presheaf construction on Pi types then I believe we do not have injectivity, then we need to specify what transp does in general, not just when the domain and codomain are constant under `\phi` ?
Is there a way to do that so that the equality in the hcomp note holds?

**Comment by simhu** [comment]
Tbh I don't remember how detailed we looked into this back then. AFAIR, no one wrote out the exact semantics for all type formers in detail somewhere, and no one proved canonicity or the like (which, as this thread suggests, might have failed without this adaption..).

But I think that there are no problems with the semantics, and they really just should follow the algorithms of the note.  One should be able to describe the semantics using the internal language.  Note that for a type `A` in context `Gamma` to have `transp` will talk about constant paths in `Gamma` (see Def. 3 of the CHM paper), and `transp` only takes such constant paths as input.

(But I also see that the typing rule of `transp` looks rather awkward, maybe an interpretation function of the syntax would be hard to define and prove correct?)

**Comment by Saizan** [comment]
> But I think that there are no problems with the semantics, and they really just should follow the algorithms of the note.

The algorithm for Pi-types in section 3.4 only covers the case when `A` and `B` are constant when `\phi` is true, though. So we do not have an algorithm for the whole of `transp^i ((x : A) -> B) \phi u` which should be given as a section of `((x : A) -> B)(i1)`, while only knowing that the whole Pi-type is constant on `\phi`.

**Comment by simhu** [comment]
To give semantics to such a system would start with proving that type formers are closed under having `transp` and `hcomp` structures.  For Pi-types we would have, say, `A : Γ -> U` and `B : (Sigma (γ : Γ) . A γ) -> U` (using the internal language of the presheaf topos) both with `transp` structures (as in Def.3 of CHM).  For `transp` of `Pi A B` we would start with `γ` in `II -> Γ` such that `φ => (i:II). γ 0 = γ i`, `w0` in `(x:A (γ 0)) -> B(x, γ 0)`, and `a1` in `A (γ 1)`.  To define `transp φ γ w0 a1` we first use transport in `A` to extend `a1` to a path `a` in `(i : II) -> A (γ i)` locally constant on `φ`.  Then one can transport `w0 (a 0)` in `B` over the path `<γ,a>` --- the usual back-and-forth argument.  This is analogous to what is done in the note, but of course not *exactly* the same as we talk about types in a context with an `i:II` on the syntax side.

Interpreting the syntax should then be similar to what is done in the CCHM setting.

**Comment by simhu** [comment]
> Interpreting the syntax should then be similar to what is done in the CCHM setting.

Hmm, maybe it is not quite clear how this interpretation works and whether the notion of transport structure in CHM is enough to interpret the syntax sketched in the note.  It might be that one has to modify the condition for transport structures on `A` from `φ => all (i:II). γ 0 = γ i` to `φ => all (i:II). A (γ 0) = A (γ i)`.  But with this modified notion it is not clear how Pi-types would work.  I really don't know/remember, someone should look into these details...

**Comment by Saizan** [comment]
Oh, indeed, it is not clear how to interpret `trans` as typed in the note or in CHM by using the transport structure as given in Def. 3. The `γ` one has to use is definitely not constant under `φ`.

It would be interesting to see a typing rule that matches Def. 3. Maybe it needs to incorporate an explicit substitution somewhere.

**Comment by andreasabel** [comment]
If #5838 is indeed an incarnation of the problem discussed here, there might be more incentive to fix it now...

**Comment by jonsterling** [comment]
> Indeed, I agree that @sattlerc 's suggestion is the best short term fix.
> 
> In the long run we should probably consider doing what @Saizan suggests and switch to the cartesian Kan operations, but that's a much bigger change. It's not clear to me if we can do this while keeping the connections and reversals on the interval while still having decidable typechecking as we would have to add "diagonal cofibrations". I have a vague recollection that @Saizan thought this was still going to be fine though?

I think we have thought about this question and although I was initially worried about the algorithm for such a system, I believe there is actually no fundamental problem. Of course, putting aside the structure on the interval, even just adding the diagonal cofibrations is likely a pretty invasive change to Agda if it was not initially designed to support them, but it would be Cool.

**Comment by Saizan** [comment]
@jonsterling with both connections and diagonal cofibrations, do you think can we still turn cofibrations into "solved form", i.e. disjunctive normal form where all equations have a variable on at least one side?

**Comment by jonsterling** [comment]
@Saizan I am not sure that this strategy will succeed, unless you replace equations by inequalities; this is actually reasonable to consider, as you can think of this in terms of a graph reachability problem.

In `cooltt`, we have an abstract type that implements a solver for the dimension theory (see http://redprl.org/cooltt/cooltt/Cubical/index.html). We use this everywhere, so most of `cooltt`'s implementation doesn't really depend on the specifics of the intension (as opposed to extension) of face propositions.

In the past we have considered adding connections for the purpose of experimenting with directed type theory.  I cannot remember how far we went but I will tag @favonia just in case they have something to add or correct.

**EDIT:** In case my main point was buried in my message, let me be a bit explicit: I think that it is good to use an architecture in which, outside of a black box module that decides face propositions, the semantic representation of your dimension elements as well as that of face propositions is not depended on. In the case of the `r=0,r=1` theory that Agda implements, the DNF where the atoms are all of the form `i=0,i=1` is a good idea, but it may be that this is not the best way to solve the problem in the presence of more sophisticated cofibrations. The first step is to ensure that you are sufficiently abstracted from these details in the main code of Agda.

**Comment by favonia** [comment]
> In the past we have considered adding connections for the purpose of experimenting with directed type theory. I cannot remember how far we went but I will tag @favonia just in case they have something to add or correct.

It's likely that, for directed type theory, the current cooltt-flavored cofibration theory extended with only `i <= j` (which would mean `i = meet(i,j)` but we don't want to introduce general meets or joins) would be sufficient for the type constructors we care about. It's trivial to add `i <= j` to cooltt, but I'd like to wait until all the details of directed type theory have been worked out on paper. There is currently an [experimental Git branch adding "directed intervals"](https://github.com/RedPRL/cooltt/pull/180) that are distinct from the existing intervals, but not much coding has been done beyond that. In sum, (1) it's possible that we only need `i <= j` and that is easy to implement, and (2) no real progress on coding has been made for some time.

**Comment by Saizan** [comment]
Using a decision procedure for "interval equality modulo assumed cofibrations" would be quite a different approach to what Agda is currently doing. Not sure how difficult it would be to integrate.

**Comment by tomjack** [comment]
In the meantime, is it worth it to "just" apply the change in simhu's updated note above? The following patch seems to fix the examples in #5838. `compHCompU DoTransp` should also be fixed, maybe? I don't know if there are any other problems...

```diff
diff --git a/src/full/Agda/TypeChecking/Primitive/Cubical.hs b/src/full/Agda/TypeChecking/Primitive/Cubical.hs
index 9ae005b62..23528c902 100644
--- a/src/full/Agda/TypeChecking/Primitive/Cubical.hs
+++ b/src/full/Agda/TypeChecking/Primitive/Cubical.hs
@@ -742,6 +742,8 @@ compGlue DoTransp psi Nothing u0 (IsFam (la, lb, bA, phi, bT, e)) tpos = do
           ug <- open $ (tunglue `apply`) . map ((setHiding Hidden) . (subst 0 iz)) $ [la, lb, bA, phi, bT, e]
           return $ \ a -> ug <@> a
         [la, lb, bA, phi, bT, e] <- mapM (\ a -> open . runNames [] $ lam "i" (const (pure $ unArg a))) [la, lb, bA, phi, bT, e]
+        let ungluei u0 i =
+              pure tunglue <#> (la <@> i) <#> (lb <@> i) <#> (bA <@> i) <#> (phi <@> i) <#> (bT <@> i) <#> (e <@> i) <@> u0
         view <- intervalView'
 
         ifM (headStop tpos (phi <@> pure io)) (return Nothing) $ Just <$> do
@@ -760,7 +762,7 @@ compGlue DoTransp psi Nothing u0 (IsFam (la, lb, bA, phi, bT, e)) tpos = do
                                              <@> psi
                                              <@> forallphi
                                              <@> ilam "o" (\ _ -> bA <@> i)
-                                             <@> ilam "o" (\ _ -> a0)
+                                             <@> ilam "o" (\ _ -> ungluei u0 i)
                                              <@> ilam "o" (\ o -> pure tEFun <#> (lb <@> i)
                                                                                <#> (la <@> i)
                                                                                <#> (bT <@> i <..> o)
```

**Comment by sattlerc** [comment]
Andreas and I just checked with a similar patch, which also seemed to fix the self-contained example in #5838.

Do you want to submit a pull request?

Don't forget to apply the same change in [a second place](https://github.com/agda/agda/blob/e2eb867be0353cec5ad1a3edf0ef209b3ed6aef4/src/full/Agda/TypeChecking/Primitive/Cubical.hs#L583) in the code (that should be refactored in the future).

**Comment by tomjack** [comment]
> Do you want to submit a pull request?

@sattlerc I will make a PR if you wish, but I think yours will be better :wink:. Also, I don't know how to fix `compHCompU DoTransp`. (I assume a similar change should be applied there?)

**Comment by andreasabel** [comment]
Ok, @tomjack, I'll make a stab at a PR.
